tool-openssl/req.cc

```diff
@@ -0,0 +1,673 @@
+  } else {
+    fprintf(stdout, "%s: ", field.short_desc);
+  }
+  fflush(stdout);
+
+  // Get input with fgets
+  if (fgets(buffer, buffer_size, stdin) == NULL) {
+    fprintf(stderr, "Error reading input\n");
+    return NULL;
+  }
+
+  // Remove newline character if present
+  size_t len = OPENSSL_strnlen(buffer, buffer_size);
+  if (len > 0 && buffer[len - 1] == '\n') {
+    buffer[len - 1] = '\0';
+    len--;
+  }
+
+  if (strcmp(buffer, ".") == 0) {
+    // Empty entry requested
+    return "";
+  }
+  if (len == 0 && field.default_value) {
+    return field.default_value;
+  }
+  if (len > 0) {
+    // Use provided input
+    return buffer;
+  }
+
+  // Empty input and no default - use empty string
+  return "";
+}
+
+static bssl::UniquePtr<X509_NAME> prompt_for_subject(
+    X509_REQ *req, bool isCSR, unsigned long chtype = MBSTRING_ASC) {
+  // Default values for subject fields
+  const ReqField subject_fields[] = {
+      {"countryName", "Country Name (2 letter code)", "AU", NID_countryName},
+      {"stateOrProvinceName", "State or Province Name (full name)",
+       "Some-State", NID_stateOrProvinceName},
+      {"localityName", "Locality Name (eg, city)", "", NID_localityName},
+      {"organizationName", "Organization Name (eg, company)",
+       "Internet Widgits Pty Ltd", NID_organizationName},
+      {"organizationalUnitName", "Organizational Unit Name (eg, section)", "",
+       NID_organizationalUnitName},
+      {"commonName", "Common Name (e.g. server FQDN or YOUR name)", "",
+       NID_commonName},
+      {"emailAddress", "Email Address", "", NID_pkcs9_emailAddress}};
+
+  // Extra attributes for CSR
+  const ReqField extra_attributes[] = {
+      {"challengePassword", "A challenge password", "",
+       NID_pkcs9_challengePassword},
+      {"unstructuredName", "An optional company name", "",
+       NID_pkcs9_unstructuredName}};
+
+  // Get the subject name from the request
+  bssl::UniquePtr<X509_NAME> subj(X509_NAME_new());
+  if (!subj) {
+    fprintf(stderr, "Error getting subject name from request\n");
+    return NULL;
+  }
+
+  // Print the instructions
+  fprintf(stdout,
+          "You are about to be asked to enter information that will be "
+          "incorporated\n");
+  fprintf(stdout, "into your certificate request.\n");
+  fprintf(stdout,
+          "What you are about to enter is what is called a Distinguished Name "
+          "or a DN.\n");
+  fprintf(stdout,
+          "There are quite a few fields but you can leave some blank\n");
+  fprintf(stdout, "For some fields there will be a default value,\n");
+  fprintf(stdout, "If you enter '.', the field will be left blank.\n");
+  fprintf(stdout, "\n");
+
+  char buffer[BUF_SIZE];
+
+  // Process each subject field
+  for (const auto &field : subject_fields) {
+    const char *value = prompt_field(field, buffer, sizeof(buffer));
+
+    if (value == NULL) {
+      return NULL;
+    }
+
+    // Only add non-empty values
+    if (value && value[0]) {
```

Comment:
NP: The second `if` condition and be simplified:
```
if (value[0]) {
...
```
... and (arguably?) clarified:
```
if (0 < OPENSSL_strnlen(value, BUF_SIZE)) {
...
```